Source/Core/DoubleEndedPriorityQueue.js

```diff
@@ -0,0 +1,406 @@
+  var that = this;
+  var maximumLength = that._maximumLength;
+  if (defined(maximumLength)) {
+    if (maximumLength === 0) {
+      return undefined;
+    } else if (that._length === maximumLength) {
+      // It's faster to access the minimum directly instead of calling the getter
+      // because it avoids the length === 0 check.
+      var minimumElement = that._array[0];
+      if (that._comparator(element, minimumElement) <= 0.0) {
+        // The element that is being inserted is less than or equal to
+        // the minimum element, so don't insert anything and exit early.
+        return element;
+      }
+      removedElement = that.removeMinimum();
+    }
+  }
+
+  var index = that._length;
+  that._array[index] = element;
+  that._length++;
+  pushUp(that, index);
+
+  return removedElement;
+};
+
+/**
+ * Removes the minimum element from the queue and returns it.
+ * If the queue is empty, the return value is undefined.
+ *
+ * @returns {*|undefined} The minimum element, or undefined if the queue is empty.
+ */
+DoubleEndedPriorityQueue.prototype.removeMinimum = function () {
+  var that = this;
+  var length = that._length;
+  if (length === 0) {
+    return undefined;
+  }
+
+  that._length--;
+
+  // The minimum element is always the root
+  var minimumElement = that._array[0];
+
+  if (length >= 2) {
+    that._array[0] = that._array[length - 1];
+    pushDown(that, 0);
+  }
+
+  // Dereference removed element
+  that._array[length - 1] = undefined;
+
+  return minimumElement;
+};
+
+/**
+ * Removes the maximum element from the queue and returns it.
+ * If the queue is empty, the return value is undefined.
+ *
+ * @returns {*|undefined} The maximum element, or undefined if the queue is empty.
+ */
+DoubleEndedPriorityQueue.prototype.removeMaximum = function () {
+  var that = this;
+  var length = that._length;
+  if (length === 0) {
+    return undefined;
+  }
+
+  that._length--;
+  var maximumElement;
+
+  // If the root has no children, the maximum is the root.
+  // If the root has one child, the maximum is the child.
+  if (length <= 2) {
+    maximumElement = that._array[length - 1];
+  } else {
+    // Otherwise, the maximum is the larger of the root's two children.
+    var maximumElementIndex = greaterThan(that, 1, 2) ? 1 : 2;
+    maximumElement = that._array[maximumElementIndex];
+
+    // Re-balance the heap
+    that._array[maximumElementIndex] = that._array[length - 1];
+    if (length >= 4) {
+      pushDown(that, maximumElementIndex);
+    }
+  }
+
+  // Dereference removed element
+  that._array[length - 1] = undefined;
+
+  return maximumElement;
+};
+
+/**
+ * Gets the minimum element in the queue.
+ * If the queue is empty, the result is undefined.
+ *
+ * @param {*|undefined} element
+ */
+
+DoubleEndedPriorityQueue.prototype.getMinimum = function () {
+  var length = this._length;
+  if (length === 0) {
+    return undefined;
+  }
+
+  // The minimum element is always the root
+  return this._array[0];
```

Comment:
This could be simplified to just `return this._array[0]`, since returning the zeroth element of an empty array is `undefined` in js

Reply:
the array could have been allocated ahead of time (if constructed with `maximumLength`), so `this._array[0]` could have junk in it